pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java

```diff
@@ -335,10 +346,88 @@ public boolean markSupported() {
     return true;
   }
 
+  /**
+   * Skips over and discards {@code n} bytes of data from this input stream.
+   *
+   * <p>Due to the "sparse" implementation of Large Objects, this class allows skipping
+   * past the "end" of the stream. Subsequent reads will continue to return {@code -1}.
+   * @see <a href="https://www.postgresql.org/docs/14/lo-implementation.html">
+   *   Large Objects: Implementation Features</a>
+   *
+   * @param n the number of bytes to be skipped.
+   * @return the actual number of bytes skipped which might be zero.
+   * @throws IOException  if an underlying driver error occurs.
+   *     In particular, this will throw if attempting to skip beyond the maximum length
+   *     of a large object, which by default is 4,398,046,509,056 bytes.
+   * @see java.io.InputStream#skip(long)
+   */
+  @Override
+  public long skip(long n) throws IOException {
+    if (n <= 0) {
+      // The spec does allow for skipping backwards, but let's not.
+      return 0;
+    }
+
+    try (ResourceLock ignore = lock.obtain()) {
+      LargeObject lo = getLo();
+      long loId = lo.getLongOID();
+
+      long targetPosition = absolutePosition + n;
+      if (targetPosition > limit || targetPosition < 0) {
+        targetPosition = limit;
+      }
+      long currentPosition = absolutePosition;
+      long skipped = targetPosition - currentPosition;
+
+      if (buffer != null && buffer.length - bufferPosition > skipped) {
+        bufferPosition += (int) skipped;
+      } else {
+        buffer = null;
+        try {
+          if (targetPosition <= Integer.MAX_VALUE) {
+            lo.seek((int) targetPosition, LargeObject.SEEK_SET);
+          } else {
+            lo.seek64(targetPosition, LargeObject.SEEK_SET);
+          }
+        } catch (SQLException e) {
+          throw new IOException(
+              GT.tr("Can not skip stream for large object {0} by {1} (currently @{2})",
+                  loId, n, currentPosition),
+              e);
+        }
+      }
+      absolutePosition = targetPosition;
+      return skipped;
+    }
+  }
+
   private LargeObject getLo() throws IOException {
-    if (lo == null) {
+    if (this.lo == null) {
       throw new IOException("BlobOutputStream is closed");
     }
+    LargeObject lo = castNonNull(this.lo);
+    assert lock.isLocked();
+
+    if (absolutePosition < 0) {
+      // Defer initialization until here so it can throw a checked exception
+      try {
+        // initialise current position for mark/reset
+        this.absolutePosition = lo.tell64();
+      } catch (SQLException e1) {
+        try {
+          // the tell64 function does not exist before PostgreSQL 9.3
+          this.absolutePosition = lo.tell();
```

Comment:
This makes the behaviour slower for the database that do not support `tell64`. We can't accept such a change.

Reply:
The last version that didn't support `tell64` was 9.2, which was EOL seven years ago. The annual PostgreSQL survey doesn't even split versions less than 10, so I cannot tell you how exactly few people that would affect.

As repeatedly explained, the only way to fix the bug is to make it slower. Though in real usage the addition time is unlikely to be noticeable compared to everything else the stream is doing.

If you have such a problem with "performance" then you should delete a load of code that prevents other bugs so that it can go a lot faster.